Ei network tutorial

ackurth-nc opened the pull request:
<!-- For questions please refer to https://lava-nc.org/developer_guide.html#how-to-contribute-to-lava or ask in a comment below -->


<!-- All pull requests require an issue https://github.com/lava-nc/lava/issues -->

<!-- Insert issue here as "Issue Number: #XXXX", example "Issue Number: #19" -->
Issue Number:  #308

<!-- Insert one sentence pr objective here, can be copied from relevant issue. -->
Add tutorial on standard network of recurrently connected excitatory and inhibitory neurons

## Pull request checklist
<!--  (Mark with "x") -->
Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [x] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [x] Build tests (`pytest`) passes locally


## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check your PR type:
<!--  (Mark one with "x") remove not chosen below -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation changes
- [x] Other (please describe): 


## Does this introduce a breaking change?
<!--  (Mark one with "x") remove not chosen below -->

- [ ] Yes
- [x] No


## Supplemental information
I also made a a diagram for the created Process. What's the workflow on adding this

awintel reviewed (approved):
Looks good. Just a few remarks:

- "how with Lava you can implement" sounds like a German has written that ;-)
- Remove "add diagram" comment or add the diagram
- "From bird's-eye view, an ..." -> "From a birds-eye..."
- "Additionaly, we here require a separation..." -> "Additionally, we require..."
- Both your EINetwork Process and SubProcModel define defaults for the Vars but different ones. This seems unnecessary.
- If I remember correctly, SubProbModels don't need a protocol specification. This gets ignored and inherited from the actually instantiated sub processes.
- Maybe say why you use both RingBuffers as source and sink... that might be surprising.
- You could consider putting the EINetwork and Ringbuffer setup in a function such that you don't have to show the same code twice in your different experiments.

The Loihi version of the tutorial should remain in intel-collab. However, the main tutorial may refer to it.

Thanks

weidel-p commented on tutorials/end_to_end/tutorial02_excitatory_inhibitory_network.ipynb:
```diff
@@ -0,0 +1,945 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsgAAAFgCAYAAACmDI9oAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAC53UlEQVR4nO2dd7wdRfn/P88555bc9N5ISAIhQOiE0DsCigqiKBYEGyroT+xYvvYS9KvYFb5YwAYoVemE3msoSQjpkBDS+80t55z5/bE7e2ZnZ7adfu/zfr2Se87u7OxsObPPPvOZ5yEhBBiGYRiGYRiGccjUuwEMwzAMwzAM00iwgcwwDMMwDMMwCmwgMwzDMAzDMIwCG8gMwzAMwzA ...
```
You could add the nice "learn more" section as in the other tutorials.

ackurth-nc replied:
Good point, addressed in 5ab2673.

weidel-p reviewed (approved):
Thanks for this nice tutorial, I think it's ready to be published. I added two minor comments complementing those by AW.

ackurth-nc commented:
The E/I tutorial has been improved now including:

- a E/I network lava process with a new, simplified interface
- multiple ProcessModels for the E/I network including implementations with rate and LIF neurons
- a preliminary mapping function of floating-point to fixed-point parameters
- a simulation of the E/I network with bit-accurate LIF and Dense ProcessModels and a comparison with the floating-point one

ackurth-nc commented on tutorials/end_to_end/tutorial02_excitatory_inhibitory_network.ipynb:
```diff
@@ -0,0 +1,945 @@
+   "metadata": {},
+   "source": [
+    "#### Controlling the network\n",
+    "One of the main control parameters we can tune for the above network is the strength of the recurrent weights. <br>\n",
```
Notebook has changed, does not appear anymore.

mathisrichter commented on tutorials/end_to_end/convert_params.py:
```diff
@@ -0,0 +1,261 @@
+
+
+# Method to convert parameters from rate to LIF
+def convert_rate_to_lif_params(**kwargs):
```
I would prefer it if you explicitly wrote down the accepted arguments (with type annotations) as arguments to the function, rather than accepting a dictionary of kwargs. That way, the user gets some hints about what he/she may pass to the function and what data types you are expecting.

```python
def convert_rate_to_lif_params(shape_exc: typing.Tuple[int, ...], ...) -> typing.Dict:
```
If this makes the calling too complicated, at least document the available arguments.

ackurth-nc replied:
Agreed, made arguments explicit it in b362062.

mathisrichter commented on tutorials/end_to_end/convert_params.py:
```diff
@@ -0,0 +1,261 @@
+from scipy.special import erf
+
+
+# Method to convert parameters from rate to LIF
```
This can be absolved into the docstring of the function. (It is a function, not a method.)

ackurth-nc replied:
Addressed in 979d63f.

mathisrichter commented on tutorials/end_to_end/convert_params.py:
```diff
@@ -0,0 +1,261 @@
+    # Define auxiliary functions for weight conversion
+    def _mean_input(weight):
+        '''
+        Calculate mean input to single neuron given mean exciatory weight
```
Typo in 'excitatory'

ackurth-nc replied:
Addressed in 979d63f.

mathisrichter commented on tutorials/end_to_end/convert_params.py:
```diff
@@ -0,0 +1,261 @@
+
+    def _std_input(weight):
+        '''
+        Calculate mean input to single neuron given mean exciatory weight
```
Typo in 'excitatory'

mathisrichter reviewed (approved):
This looks very good and polished. I gave some detailed feedback on some portions of code though, somewhat minor points.

There are some more general points that require attention though.
- None of the code seems to be unit tested. Code that does not have automated unit tests should usually not pass code reviews. I would expect unit tests at least for the function that converts rate-code parameters to LIF parameters. Make sure that you think through any qualitatively different cases of input and write tests for those. What happens, for instance, if the user does not supply `shape_exc` or misspells it? I would usually mark my review as 'Request changes' for now but will approve it given the upcoming release. If you can add in tests, that would be great. For future development, please consider adopting test-driven development as a habit.
- The code is missing type annotations throughout. Please consider getting into the habit of adding type annotations to any code you are writing.
- Following the Lava developer guidelines, all comments should be sentences that begin with a capital letter and end in a full stop.

mathisrichter commented on tutorials/end_to_end/tutorial02_excitatory_inhibitory_network.ipynb:
```diff
@@ -0,0 +1,1813 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAb8AAAE+CAYAAAAd/x4PAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAEAAElEQVR4nOy9dbikZ33//3pk3OXMHPez7pbdbFyJEhJCCMVboFhpgRqlv0KdtpT2S5FSoMWCREggQlw22exu1n2Pu8wZ95lH7t8fs9kkZJNsko0A87qu55ojj9xj9/v53B+ThBDUqVOnTp06v0vIb/QA6tSpU6dOndebuvjVqVOnTp3fOeriV6dOnTp1fueoi1+dOnXq1Pmdoy5+der ...
```
This does not show the correct section of code I wanted to highlight. I was referring to this section in the Jupyter notebook.
```python
    def convert_rate_to_lif_params(**kwargs):
        '''Convert rate parameters to LIF parameters.
        The mapping is based on A unified view on weakly correlated recurrent network, Grytskyy et al. 2013
        
        Parameters
        ----------
        kwargs : dict
            Parameter dictionary for rate network
        
        Returns
        -------
        lif_network_dict : dict
            Parameter dictionary for LIF network
        '''       
    
        return convert_rate_to_lif_params(**kwargs)
```

ackurth-nc replied:
I agree, this is unnecessary. Addressed in 2eddfcd.

mathisrichter commented on tutorials/end_to_end/tutorial02_excitatory_inhibitory_network.ipynb:
```diff
@@ -0,0 +1,1813 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAb8AAAE+CAYAAAAd/x4PAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAEAAElEQVR4nOy9dbikZ33//3pk3OXMHPez7pbdbFyJEhJCCMVboFhpgRqlv0KdtpT2S5FSoMWCREggQlw22exu1n2Pu8wZ95lH7t8fs9kkZJNsko0A87qu55ojj9xj9/v53B+ThBDUqVOnTp06v0vIb/QA6tSpU6dOndebuvjVqVOnTp3fOeriV6dOnTp1fueoi1+dOnXq1Pmdoy5+der ...
```
Same as above... I was referring to the raster plot function.
```python
def raster_plot(spks, stride=6, fig=None, color='b', alpha=1):
```

ackurth-nc replied:
I think it would not be wise to make the tutorial be dependent on another PR.
I volunteer to do this in after the completion of this PR.
See  #362

ackurth-nc commented on tutorials/end_to_end/convert_params.py:
```diff
@@ -0,0 +1,261 @@
+import numpy as np
```
Addressed in 979d63f.

ackurth-nc commented on tutorials/end_to_end/tutorial02_excitatory_inhibitory_network.ipynb:
```diff
@@ -0,0 +1,1813 @@
+   "id": "7febf4a0",
+   "metadata": {},
+   "source": [
+    "#### Visualizing the acitvity\n",
```
Addressed in 979d63f.

tim-shea reviewed (approved):
Comments on tutorial:
The intro text at top is too long. The bold "First aim", "secondly", and "finally" can all be cut. Recommend changing "bird's eye view" to "high level". Recommend simplifying/shortening "E/I Network".

All imports should be collected at the top unless they are conditional or you are going to spend some time explaining something about them.

Recommend factoring generate_gaussian_weights out, since this is a fundamental component of what an E/I net is, it should be properly introduced and described.

"Defining the parameters..." text is too long.

The text below "Visualizing the network" does not match my impression from the plot. Please try to explain what I am seeing in the plot with an outsider's perspective.

The covariance analysis feels too academic and unmotivated. For both this, and the previous plot, maybe show two side-by-side plots such that you can illustrate to the developer what is different or what question could be answered through this analysis.

"chaotical" > "chaotic"

Note that the chaotic network also seems to be drifting at a macro-level. If you run this for longer, you may see more interesting behavior that could be analysed. In addition, both network state plots could be improved by plotting fewer overlapping neuron states. E.g. plot just the first 25 dimensions.

CustomRunConfigFloat is not explained and the "if isinstance" pattern is an anti-pattern. Can't you accomplish this selection using tags without overriding select?

The text in "Self-balancing of the network" is too academic.

In the small vs high weights plot, one of these labels should be changed (either small vs large, or low vs high).

Typo in "Runnig a process model"

Overall and most importantly, each individual stage of analysis seems technically fine, but there doesn't seem to be any clear, driving purpose to this tutorial. I think it would strongly benefit from thinking about "what is the one main thing we are trying to teach people with this tutorial?" and then structuring the stages to progressively work towards that goal.

ackurth-nc commented on tutorials/end_to_end/convert_params.py:
```diff
@@ -0,0 +1,261 @@
+
+
+# Method to convert parameters from rate to LIF
+def convert_rate_to_lif_params(**kwargs):
```
Addressed in c53222a.

ackurth-nc commented:
Thanks, Tim, for the thorough review. 

I addressed some of the remarks in the commit c53222a.
> Comments on tutorial: The intro text at top is too long. The bold "First aim", "secondly", and "finally" can all be cut. Recommend changing "bird's eye view" to "high level". Recommend simplifying/shortening "E/I Network".
> 
Followed your recommendations.
> All imports should be collected at the top unless they are conditional or you are going to spend some time explaining something about them.
> 
I of course agree for a single script. Is this also best practice in a tutorial? In the other tutorials it is done as in this.
> Recommend factoring generate_gaussian_weights out, since this is a fundamental component of what an E/I net is, it should be properly introduced and described.
> 
Done.
> "Defining the parameters..." text is too long.
> 
Adapted text.
> The text below "Visualizing the network" does not match my impression from the plot. Please try to explain what I am seeing in the plot with an outsider's perspective.
> 
Adapted text.
> The covariance analysis feels too academic and unmotivated. For both this, and the previous plot, maybe show two side-by-side plots such that you can illustrate to the developer what is different or what question could be answered through this analysis.
> 
I adapted the text.
> "chaotical" > "chaotic"
> 
Done.
> Note that the chaotic network also seems to be drifting at a macro-level. If you run this for longer, you may see more interesting behavior that could be analysed. In addition, both network state plots could be improved by plotting fewer overlapping neuron states. E.g. plot just the first 25 dimensions.
> 
Changed the plotting to plot fewer neurons.
> CustomRunConfigFloat is not explained and the "if isinstance" pattern is an anti-pattern. Can't you accomplish this selection using tags without overriding select?
> 
I wrote a sentence for the CustomRunConfig. Regarding you question: I would not know how to do this. When I select a tag, say as in the tutorial ``` 'lif_neurons' ```  for the ProcessModel of the EINetwork and the RingBuffer has no implementation for this tag, we get problems I think. But maybe I am mistaken! Furthermore, I want to also specifically choose my LIF and Dense ProcessModel (for the bit-accurate runs later). The code snippet for the CustomRunConfig comes from [here](https://github.com/lava-nc/lava-dl/blob/main/tutorials/lava/lib/dl/netx/pilotnet_snn/run.ipynb).
> The text in "Self-balancing of the network" is too academic.
> 
Adapted the text a bit. If you think it does not fit at all I can also remove it.
> In the small vs high weights plot, one of these labels should be changed (either small vs large, or low vs high).
> 
Done.
> Typo in "Runnig a process model"
> 
Done.
> Overall and most importantly, each individual stage of analysis seems technically fine, but there doesn't seem to be any clear, driving purpose to this tutorial. I think it would strongly benefit from thinking about "what is the one main thing we are trying to teach people with this tutorial?" and then structuring the stages to progressively work towards that goal.

I see what you mean. The main purpose of the tutorial, to my understanding, is showcasing of how to work with different process models that implement different behavior but still share computational similarities. To see this, I need the analysis,
For this reason there is also a unified interface for the rate and LIF network as well as the conversion functionality.
Does this make it a bit clearer?

awintel reviewed (commented):
Looks good!

Names are inconsistent:
    def __init__(self, proc):
        super().__init__(proc_params=proc)
It's either a proc or a proc_param but a proc shouldn't turn into proc_params.

Typo:
hot to define and select

Inconsistency:
Some single line comments end with a full stop, other don't.

Why do we not quantitatively compare the rate, float-LIF and fixed-LIF models?

ackurth-nc commented:
Thanks for the review, I addressed your points in  f2793b7.

> Names are inconsistent: def **init**(self, proc): super().**init**(proc_params=proc) It's either a proc or a proc_param but a proc shouldn't turn into proc_params.
> 
Great catch! I made it consistent.
> Typo: hot to define and select
> 
Fixed typo.
> Inconsistency: Some single line comments end with a full stop, other don't.
> 
Made it consistent.
> Why do we not quantitatively compare the rate, float-LIF and fixed-LIF models?
It is not directly straightforward because the rate in a LIF network the rate is a priori positive but in the rate network the state can also attain negative values. In a way it's more like a temporal averaged rate.
I added a rate comparison where one can see that the behavior actually is very similar when averaged correctly.
If you think it does not fit in this form, we can also kick it out again.
